nx/lib/nx.ex

```diff
@@ -13293,83 +13293,97 @@ defmodule Nx do
+    end
 
-    canonical_vectorized_axes = calculate_canonical_vectorized_axes([tensor, indices])
-    target_vectorized_axes = Keyword.keys(canonical_vectorized_axes)
+    opts = keyword!(opts, axis: 0)
 
-    offset = length(canonical_vectorized_axes)
+    tensor = to_tensor(tensor)
+    indices = to_tensor(indices)
 
-    indices =
-      [indices]
-      |> do_reshape_vectors_devectorize(
-        canonical_vectorized_axes,
-        offset,
-        false
+    axis =
+      Nx.Shape.normalize_axis(
+        tensor.shape,
+        opts[:axis],
+        tensor.names
       )
-      |> hd()
-      |> vectorize(target_vectorized_axes)
-
-    unless Nx.Type.integer?(indices.type) do
-      raise ArgumentError, "indices must be an integer tensor, got #{inspect(indices.type)}"
-    end
-
-    opts = keyword!(opts, axis: 0)
-    axis = Nx.Shape.normalize_axis(tensor.shape, opts[:axis], tensor.names)
 
-    {shape, names, indices} =
-      if offset != 0 do
-        {shape, names} =
-          Nx.Shape.take(tensor.shape, tensor.names, indices.shape, indices.names, axis)
-
-        # because take/3 will insert the index shape into the tensor axis which we are
-        # taking from, we need to ensure that we squeeze out all common vectorized
-        # axes, as well as the 1-sized vectorized axes from indices
-        {[], indices_slice_lengths_rev, indices_squeeze_axes_rev, _} =
-          Enum.reduce(indices.vectorized_axes, {tensor.vectorized_axes, [], [], 0}, fn
-            {_name, s1}, {[], slice_acc, squeeze_acc, axis} ->
-              {[], [s1 | slice_acc], squeeze_acc, axis + 1}
-
-            {name, s1}, {other_axes, slice_acc, squeeze_acc, axis} ->
-              {slice, squeeze_acc, other_axes} =
-                case List.keytake(other_axes, name, 0) do
-                  {{^name, s2}, other_axes} when s2 >= s1 ->
-                    {1, [axis | squeeze_acc], other_axes}
-
-                  _ ->
-                    {s1, squeeze_acc, other_axes}
-                end
+    {inner_shape, inner_names} =
+      Nx.Shape.take(
+        tensor.shape,
+        tensor.names,
+        indices.shape,
+        indices.names,
+        axis
+      )
 
-              {other_axes, [slice | slice_acc], squeeze_acc, axis + 1}
-          end)
+    if tensor.vectorized_axes != [] or indices.vectorized_axes != [] do
+      axes_range = axes(tensor)
 
-        devec_indices = devectorize(indices, keep_names: false)
-        starts = List.duplicate(0, rank(devec_indices))
-        lengths = Enum.reverse(indices_slice_lengths_rev) ++ Tuple.to_list(indices.shape)
+      indices = if rank(indices) == 0, do: new_axis(indices, 0), else: indices
 
-        indices =
-          devec_indices
-          |> slice(starts, lengths)
-          |> squeeze(axes: Enum.reverse(indices_squeeze_axes_rev))
+      indices_shape =
+        axes_range
+        |> Enum.flat_map(fn
+          ^axis -> Tuple.to_list(indices.shape)
```

Comment:
Perhaps we could use `Tuple.product(indices.shape)` and remove `axis_offset`, since it's all flattened out anyway?